File: mobile/ios/Security/AntiTamperingManager.swift
import Foundation
import Security
import CommonCrypto

class AntiTamperingManager {
    static let shared = AntiTamperingManager()
    
    private let expectedChecksums: [String: String]
    private let criticalFunctions: [String]
    private var validationResults: [String: Bool] = [:]
    
    private init() {
        self.expectedChecksums = AntiTamperingManager.loadExpectedChecksums()
        self.criticalFunctions = [
            "performSecurityChecks",
            "validateUserInput",
            "encryptSensitiveData",
            "decryptSensitiveData",
            "validateCertificate",
            "authenticateUser"
        ]
    }
    
    // MARK: - Main Validation Methods
    
    /// Проверка целостности приложения
    func validateAppIntegrity() -> TamperingDetectionResult {
        var detectionMethods: [TamperingCheck: Bool] = [:]
        var riskLevel: TamperingRiskLevel = .low
        
        // 1. Проверка подписи приложения
        detectionMethods[.appSignature] = validateAppSignature()
        
        // 2. Проверка целостности критических функций
        detectionMethods[.criticalFunctions] = validateCriticalFunctions()
        
        // 3. Проверка целостности ресурсов
        detectionMethods[.resources] = validateResources()
        
        // 4. Проверка целостности бинарного файла
        detectionMethods[.binaryIntegrity] = validateBinaryIntegrity()
        
        // 5. Проверка контрольных сумм
        detectionMethods[.checksums] = validateChecksums()
        
        // 6. Проверка динамической библиотеки
        detectionMethods[.dynamicLibraries] = validateDynamicLibraries()
        
        // 7. Проверка конфигурации
        detectionMethods[.configuration] = validateConfiguration()
        
        // Определяем уровень риска
        let violations = detectionMethods.values.filter { !$0 }.count
        riskLevel = determineTamperingRiskLevel(violations: violations)
        
        let isTampered = violations > 0
        
        return TamperingDetectionResult(
            isTampered: isTampered,
            riskLevel: riskLevel,
            detectionMethods: detectionMethods,
            timestamp: Date()
        )
    }
    
    // MARK: - Validation Methods
    
    // 1. Проверка подписи приложения
    private func validateAppSignature() -> Bool {
        guard let bundlePath = Bundle.main.bundlePath.cString(using: .utf8) else {
            print("❌ Anti-Tampering: Failed to get bundle path")
            return false
        }
        
        var staticCode: SecStaticCode?
        let status = SecStaticCodeCreateWithPath(
            URL(fileURLWithPath: Bundle.main.bundlePath) as CFURL,
            [],
            &staticCode
        )
        
        guard status == errSecSuccess, let code = staticCode else {
            print("❌ Anti-Tampering: Failed to create static code")
            return false
        }
        
        let validationResult = SecStaticCodeCheckValidity(code, [], nil)
        let isValid = validationResult == errSecSuccess
        
        if !isValid {
            print("🚨 Anti-Tampering: App signature validation failed")
        } else {
            print("✅ Anti-Tampering: App signature valid")
        }
        
        return isValid
    }
    
    // 2. Проверка целостности критических функций
    private func validateCriticalFunctions() -> Bool {
        var allValid = true
        
        for functionName in criticalFunctions {
            if let symbol = dlsym(dlopen(nil, RTLD_NOW), functionName) {
                // Проверяем, что функция не была изменена
                let checksum = calculateChecksum(for: symbol)
                
                if let expectedChecksum = expectedChecksums[functionName] {
                    if checksum != expectedChecksum {
                        print("🚨 Anti-Tampering: Function \(functionName) modified")
                        allValid = false
                    } else {
                        print("✅ Anti-Tampering: Function \(functionName) intact")
                    }
                }
            } else {
                print("⚠️ Anti-Tampering: Function \(functionName) not found")
            }
        }
        
        return allValid
    }
    
    // 3. Проверка целостности ресурсов
    private func validateResources() -> Bool {
        let criticalResources = [
            "Info.plist",
            "Base.lproj",
            "Assets.car"
        ]
        
        var allValid = true
        
        for resource in criticalResources {
            if let resourcePath = Bundle.main.path(forResource: resource, ofType: nil) {
                let checksum = calculateFileChecksum(path: resourcePath)
                
                if let expectedChecksum = expectedChecksums[resource] {
                    if checksum != expectedChecksum {
                        print("🚨 Anti-Tampering: Resource \(resource) modified")
                        allValid = false
                    } else {
                        print("✅ Anti-Tampering: Resource \(resource) intact")
                    }
                }
            } else {
                print("⚠️ Anti-Tampering: Resource \(resource) not found")
            }
        }
        
        return allValid
    }
    
    // 4. Проверка целостности бинарного файла
    private func validateBinaryIntegrity() -> Bool {
        guard let executablePath = Bundle.main.executablePath else {
            print("❌ Anti-Tampering: Failed to get executable path")
            return false
        }
        
        let checksum = calculateFileChecksum(path: executablePath)
        
        if let expectedChecksum = expectedChecksums["executable"] {
            let isValid = checksum == expectedChecksum
            
            if !isValid {
                print("🚨 Anti-Tampering: Executable modified")
            } else {
                print("✅ Anti-Tampering: Executable intact")
            }
            
            return isValid
        }
        
        return true
    }
    
    // 5. Проверка контрольных сумм
    private func validateChecksums() -> Bool {
        var allValid = true
        
        for (file, expectedChecksum) in expectedChecksums {
            if let filePath = Bundle.main.path(forResource: file, ofType: nil) {
                let actualChecksum = calculateFileChecksum(path: filePath)
                
                if actualChecksum != expectedChecksum {
                    print("🚨 Anti-Tampering: Checksum mismatch for \(file)")
                    allValid = false
                } else {
                    print("✅ Anti-Tampering: Checksum valid for \(file)")
                }
            }
        }
        
        return allValid
    }
    
    // 6. Проверка динамических библиотек
    private func validateDynamicLibraries() -> Bool {
        let suspiciousLibraries = [
            "MobileSubstrate",
            "libsubstrate",
            "libhooker",
            "libsubstitute"
        ]
        
        for library in suspiciousLibraries {
            if dlopen(library, RTLD_NOW) != nil {
                print("🚨 Anti-Tampering: Suspicious library detected: \(library)")
                return false
            }
        }
        
        print("✅ Anti-Tampering: No suspicious libraries")
        return true
    }
    
    // 7. Проверка конфигурации
    private func validateConfiguration() -> Bool {
        guard let infoDictionary = Bundle.main.infoDictionary else {
            print("❌ Anti-Tampering: Failed to get info dictionary")
            return false
        }
        
        // Проверяем критические ключи конфигурации
        let criticalKeys = [
            "CFBundleIdentifier",
            "CFBundleVersion",
            "CFBundleShortVersionString"
        ]
        
        for key in criticalKeys {
            if infoDictionary[key] == nil {
                print("🚨 Anti-Tampering: Missing configuration key: \(key)")
                return false
            }
        }
        
        print("✅ Anti-Tampering: Configuration valid")
        return true
    }
    
    // MARK: - Helper Methods
    
    private static func loadExpectedChecksums() -> [String: String] {
        // В реальном приложении эти значения были бы зашифрованы и обфусцированы
        return [
            "executable": "abc123def456",
            "Info.plist": "def456ghi789",
            "Base.lproj": "ghi789jkl012",
            "performSecurityChecks": "jkl012mno345",
            "validateUserInput": "mno345pqr678"
        ]
    }
    
    private func calculateChecksum(for symbol: UnsafeRawPointer) -> String {
        // Упрощенная версия - в реальном приложении здесь был бы SHA-256
        let address = Int(bitPattern: symbol)
        return String(format: "%016x", address)
    }
    
    private func calculateFileChecksum(path: String) -> String {
        guard let data = FileManager.default.contents(atPath: path) else {
            return ""
        }
        
        var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
        data.withUnsafeBytes {
            _ = CC_SHA256($0.baseAddress, CC_LONG(data.count), &hash)
        }
        
        return hash.map { String(format: "%02x", $0) }.joined()
    }
    
    private func determineTamperingRiskLevel(violations: Int) -> TamperingRiskLevel {
        switch violations {
        case 0:
            return .low
        case 1...2:
            return .medium
        case 3...4:
            return .high
        default:
            return .critical
        }
    }
    
    // MARK: - Public Methods
    
    /// Получить детальный отчет
    func getDetailedReport() -> String {
        let result = validateAppIntegrity()
        
        var report = "🔒 Anti-Tampering Report\n"
        report += "========================\n"
        report += "Status: \(result.isTampered ? "🚨 TAMPERED" : "✅ INTACT")\n"
        report += "Risk Level: \(result.riskLevel.rawValue)\n"
        report += "Timestamp: \(result.timestamp)\n\n"
        
        report += "Validation Checks:\n"
        for (check, isValid) in result.detectionMethods {
            report += "- \(check.rawValue): \(isValid ? "✅ VALID" : "🚨 INVALID")\n"
        }
        
        return report
    }
    
    /// Быстрая проверка целостности
    func quickIntegrityCheck() -> Bool {
        let result = validateAppIntegrity()
        return !result.isTampered
    }
}

// MARK: - Data Models

struct TamperingDetectionResult {
    let isTampered: Bool
    let riskLevel: TamperingRiskLevel
    let detectionMethods: [TamperingCheck: Bool]
    let timestamp: Date
}

enum TamperingRiskLevel: String, CaseIterable {
    case low = "Low"
    case medium = "Medium"
    case high = "High"
    case critical = "Critical"
}

enum TamperingCheck: String, CaseIterable {
    case appSignature = "App Signature"
    case criticalFunctions = "Critical Functions"
    case resources = "Resources"
    case binaryIntegrity = "Binary Integrity"
    case checksums = "Checksums"
    case dynamicLibraries = "Dynamic Libraries"
    case configuration = "Configuration"
}

File: mobile/android/Security/AntiTamperingManager.kt
package com.aladdin.mobile.security

import android.content.Context
import android.content.pm.PackageManager
import android.util.Log
import java.io.File
import java.security.MessageDigest
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AntiTamperingManager @Inject constructor(
    private val context: Context
) {
    
    private val expectedChecksums: Map<String, String>
    private val criticalClasses: List<String>
    
    init {
        expectedChecksums = loadExpectedChecksums()
        criticalClasses = listOf(
            "com.aladdin.mobile.security.CertificatePinningManager",
            "com.aladdin.mobile.security.RootDetector",
            "com.aladdin.mobile.security.RASPManager",
            "com.aladdin.mobile.network.ALADDINNetworkManager"
        )
    }
    
    // MARK: - Main Validation Methods
    
    /**
     * Проверка целостности приложения
     */
    fun validateAppIntegrity(): TamperingDetectionResult {
        var detectionMethods: MutableMap<TamperingCheck, Boolean> = mutableMapOf()
        var riskLevel: TamperingRiskLevel = TamperingRiskLevel.LOW
        
        // 1. Проверка подписи приложения
        detectionMethods[TamperingCheck.APP_SIGNATURE] = validateAppSignature()
        
        // 2. Проверка целостности критических классов
        detectionMethods[TamperingCheck.CRITICAL_CLASSES] = validateCriticalClasses()
        
        // 3. Проверка целостности ресурсов
        detectionMethods[TamperingCheck.RESOURCES] = validateResources()
        
        // 4. Проверка целостности APK
        detectionMethods[TamperingCheck.APK_INTEGRITY] = validateApkIntegrity()
        
        // 5. Проверка контрольных сумм
        detectionMethods[TamperingCheck.CHECKSUMS] = validateChecksums()
        
        // 6. Проверка установщика
        detectionMethods[TamperingCheck.INSTALLER] = validateInstaller()
        
        // 7. Проверка манифеста
        detectionMethods[TamperingCheck.MANIFEST] = validateManifest()
        
        // Определяем уровень риска
        val violations = detectionMethods.values.count { !it }
        riskLevel = determineTamperingRiskLevel(violations)
        
        val isTampered = violations > 0
        
        return TamperingDetectionResult(
            isTampered = isTampered,
            riskLevel = riskLevel,
            detectionMethods = detectionMethods,
            timestamp = System.currentTimeMillis()
        )
    }
    
    // MARK: - Validation Methods
    
    // 1. Проверка подписи приложения
    private fun validateAppSignature(): Boolean {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(
                context.packageName,
                PackageManager.GET_SIGNATURES
            )
            
            val signatures = packageInfo.signatures
            if (signatures.isNotEmpty()) {
                val signature = signatures[0]
                val md = MessageDigest.getInstance("SHA-256")
                md.update(signature.toByteArray())
                val hash = md.digest()
                val hashString = hash.joinToString("") { "%02x".format(it) }
                
                val expectedSignature = expectedChecksums["app_signature"]
                val isValid = expectedSignature == null || hashString == expectedSignature
                
                if (isValid) {
                    Log.i("AntiTampering", "App signature valid")
                } else {
                    Log.w("AntiTampering", "App signature invalid")
                }
                
                isValid
            } else {
                Log.w("AntiTampering", "No signatures found")
                false
            }
        } catch (e: Exception) {
            Log.e("AntiTampering", "Error validating app signature", e)
            false
        }
    }
    
    // 2. Проверка целостности критических классов
    private fun validateCriticalClasses(): Boolean {
        var allValid = true
        
        for (className in criticalClasses) {
            try {
                val clazz = Class.forName(className)
                val checksum = calculateClassChecksum(clazz)
                
                val expectedChecksum = expectedChecksums[className]
                if (expectedChecksum != null && checksum != expectedChecksum) {
                    Log.w("AntiTampering", "Class $className modified")
                    allValid = false
                } else {
                    Log.i("AntiTampering", "Class $className intact")
                }
            } catch (e: ClassNotFoundException) {
                Log.w("AntiTampering", "Class $className not found")
                allValid = false
            }
        }
        
        return allValid
    }
    
    // 3. Проверка целостности ресурсов
    private fun validateResources(): Boolean {
        val criticalResources = listOf(
            "AndroidManifest.xml",
            "classes.dex",
            "resources.arsc"
        )
        
        var allValid = true
        
        for (resource in criticalResources) {
            val checksum = calculateResourceChecksum(resource)
            val expectedChecksum = expectedChecksums[resource]
            
            if (expectedChecksum != null && checksum != expectedChecksum) {
                Log.w("AntiTampering", "Resource $resource modified")
                allValid = false
            } else {
                Log.i("AntiTampering", "Resource $resource intact")
            }
        }
        
        return allValid
    }
    
    // 4. Проверка целостности APK
    private fun validateApkIntegrity(): Boolean {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(
                context.packageName,
                0
            )
            
            val apkPath = packageInfo.applicationInfo.sourceDir
            val apkFile = File(apkPath)
            
            if (!apkFile.exists()) {
                Log.w("AntiTampering", "APK file not found")
                return false
            }
            
            val checksum = calculateFileChecksum(apkFile)
            val expectedChecksum = expectedChecksums["apk"]
            
            val isValid = expectedChecksum == null || checksum == expectedChecksum
            
            if (isValid) {
                Log.i("AntiTampering", "APK integrity valid")
            } else {
                Log.w("AntiTampering", "APK integrity invalid")
            }
            
            isValid
        } catch (e: Exception) {
            Log.e("AntiTampering", "Error validating APK integrity", e)
            false
        }
    }
    
    // 5. Проверка контрольных сумм
    private fun validateChecksums(): Boolean {
        var allValid = true
        
        for ((file, expectedChecksum) in expectedChecksums) {
            // Проверяем существование файла и его контрольную сумму
            val actualChecksum = when {
                file.startsWith("com.aladdin") -> calculateClassChecksum(Class.forName(file))
                file.endsWith(".xml") || file.endsWith(".dex") -> calculateResourceChecksum(file)
                else -> continue
            }
            
            if (actualChecksum != expectedChecksum) {
                Log.w("AntiTampering", "Checksum mismatch for $file")
                allValid = false
            } else {
                Log.i("AntiTampering", "Checksum valid for $file")
            }
        }
        
        return allValid
    }
    
    // 6. Проверка установщика
    private fun validateInstaller(): Boolean {
        val installerPackageName = context.packageManager.getInstallerPackageName(context.packageName)
        
        val validInstallers = listOf(
            "com.android.vending", // Google Play
            "com.google.android.feedback", // Google Play (альтернативный)
            null // Разрешаем установку через ADB в debug режиме
        )
        
        val isValid = validInstallers.contains(installerPackageName)
        
        if (isValid) {
            Log.i("AntiTampering", "Installer valid: $installerPackageName")
        } else {
            Log.w("AntiTampering", "Installer invalid: $installerPackageName")
        }
        
        return isValid
    }
    
    // 7. Проверка манифеста
    private fun validateManifest(): Boolean {
        return try {
            val packageInfo = context.packageManager.getPackageInfo(
                context.packageName,
                PackageManager.GET_PERMISSIONS
            )
            
            // Проверяем критические разрешения
            val criticalPermissions = listOf(
                android.Manifest.permission.INTERNET,
                android.Manifest.permission.ACCESS_NETWORK_STATE
            )
            
            val requestedPermissions = packageInfo.requestedPermissions?.toList() ?: emptyList()
            
            var allValid = true
            for (permission in criticalPermissions) {
                if (!requestedPermissions.contains(permission)) {
                    Log.w("AntiTampering", "Missing permission: $permission")
                    allValid = false
                }
            }
            
            if (allValid) {
                Log.i("AntiTampering", "Manifest valid")
            }
            
            allValid
        } catch (e: Exception) {
            Log.e("AntiTampering", "Error validating manifest", e)
            false
        }
    }
    
    // MARK: - Helper Methods
    
    private fun loadExpectedChecksums(): Map<String, String> {
        // В реальном приложении эти значения были бы зашифрованы и обфусцированы
        return mapOf(
            "app_signature" to "abc123def456ghi789jkl012mno345pqr678stu901vwx234yz",
            "apk" to "def456ghi789jkl012mno345pqr678stu901vwx234yz567abc",
            "AndroidManifest.xml" to "ghi789jkl012mno345pqr678stu901",
            "classes.dex" to "jkl012mno345pqr678stu901vwx234",
            "resources.arsc" to "mno345pqr678stu901vwx234yz567"
        )
    }
    
    private fun calculateClassChecksum(clazz: Class<*>): String {
        return try {
            val classLoader = clazz.classLoader
            val className = clazz.name.replace(".", "/") + ".class"
            val inputStream = classLoader?.getResourceAsStream(className)
            
            val md = MessageDigest.getInstance("SHA-256")
            val buffer = ByteArray(8192)
            var bytesRead: Int
            
            while (inputStream?.read(buffer).also { bytesRead = it ?: -1 } != -1) {
                md.update(buffer, 0, bytesRead)
            }
            
            inputStream?.close()
            
            val hash = md.digest()
            hash.joinToString("") { "%02x".format(it) }
        } catch (e: Exception) {
            Log.e("AntiTampering", "Error calculating class checksum", e)
            ""
        }
    }
    
    private fun calculateResourceChecksum(resourceName: String): String {
        return try {
            val assetManager = context.assets
            val inputStream = assetManager.open(resourceName)
            
            val md = MessageDigest.getInstance("SHA-256")
            val buffer = ByteArray(8192)
            var bytesRead: Int
            
            while (inputStream.read(buffer).also { bytesRead = it } != -1) {
                md.update(buffer, 0, bytesRead)
            }
            
            inputStream.close()
            
            val hash = md.digest()
            hash.joinToString("") { "%02x".format(it) }
        } catch (e: Exception) {
            Log.e("AntiTampering", "Error calculating resource checksum", e)
            ""
        }
    }
    
    private fun calculateFileChecksum(file: File): String {
        return try {
            val md = MessageDigest.getInstance("SHA-256")
            val inputStream = file.inputStream()
            val buffer = ByteArray(8192)
            var bytesRead: Int
            
            while (inputStream.read(buffer).also { bytesRead = it } != -1) {
                md.update(buffer, 0, bytesRead)
            }
            
            inputStream.close()
            
            val hash = md.digest()
            hash.joinToString("") { "%02x".format(it) }
        } catch (e: Exception) {
            Log.e("AntiTampering", "Error calculating file checksum", e)
            ""
        }
    }
    
    private fun determineTamperingRiskLevel(violations: Int): TamperingRiskLevel {
        return when (violations) {
            0 -> TamperingRiskLevel.LOW
            1, 2 -> TamperingRiskLevel.MEDIUM
            3, 4 -> TamperingRiskLevel.HIGH
            else -> TamperingRiskLevel.CRITICAL
        }
    }
    
    // MARK: - Public Methods
    
    /**
     * Получить детальный отчет
     */
    fun getDetailedReport(): String {
        val result = validateAppIntegrity()
        
        var report = "🔒 Anti-Tampering Report\n"
        report += "========================\n"
        report += "Status: ${if (result.isTampered) "🚨 TAMPERED" else "✅ INTACT"}\n"
        report += "Risk Level: ${result.riskLevel.name}\n"
        report += "Timestamp: ${java.util.Date(result.timestamp)}\n\n"
        
        report += "Validation Checks:\n"
        for ((check, isValid) in result.detectionMethods) {
            report += "- ${check.name}: ${if (isValid) "✅ VALID" else "🚨 INVALID"}\n"
        }
        
        return report
    }
    
    /**
     * Быстрая проверка целостности
     */
    fun quickIntegrityCheck(): Boolean {
        val result = validateAppIntegrity()
        return !result.isTampered
    }
}

// MARK: - Data Models

data class TamperingDetectionResult(
    val isTampered: Boolean,
    val riskLevel: TamperingRiskLevel,
    val detectionMethods: Map<TamperingCheck, Boolean>,
    val timestamp: Long
)

enum class TamperingRiskLevel {
    LOW, MEDIUM, HIGH, CRITICAL
}

enum class TamperingCheck {
    APP_SIGNATURE,
    CRITICAL_CLASSES,
    RESOURCES,
    APK_INTEGRITY,
    CHECKSUMS,
    INSTALLER,
    MANIFEST
}

